File: server/internal/tiler/pyramid.go
package tiler

func LevelPixels(size int, z int) int {
	if size <= 0 {
		return 0
	}
	if z >= 62 {
		return 1
	}
	step := 1 << uint(z)
	if step >= size {
		return 1
	}
	return (size + step - 1) >> uint(z)
}
func LevelTiles(size int, tileSize int, z int) int {
	if tileSize < 1 {
		return 0
	}
	pixels := LevelPixels(size, z)
	return (pixels + tileSize - 1) / tileSize
}
func LevelTileSize(size int, tileSize int, z int, x int) int {
	if tileSize < 1 || x < 0 || x >= LevelTiles(size, tileSize, z) {
		return 0
	}
	if remainder := LevelPixels(size, z) - x*tileSize; remainder < tileSize {
		return remainder
	}
	return tileSize
}
func MaxZoom(width int, height int, tileSize int) int {
	if tileSize < 1 {
		return 0
	}
	z := 0
	for LevelPixels(width, z) > tileSize || LevelPixels(height, z) > tileSize {
		z++
	}
	return z
}

File: server/js/room/render/pyramid.ts






















export function levelPixels(size: number, z: number): number {
	if (size <= 0) {
		return 0;
	}
	if (z >= 31) {
		return 1;
	}

	const step = 1 << z;
	if (step >= size) {
		return 1;
	}

	return (size + step - 1) >> z;
}



export function levelTiles(size: number, tileSize: number, z: number): number {
	if (tileSize < 1) {
		return 0;
	}

	return Math.ceil(levelPixels(size, z) / tileSize);
}








export function levelTileSize(size: number, tileSize: number, z: number, x: number): number {
	if (tileSize < 1 || x < 0 || x >= levelTiles(size, tileSize, z)) {
		return 0;
	}

	const remainder = levelPixels(size, z) - x * tileSize;

	return remainder < tileSize ? remainder : tileSize;
}




export function maxZoom(width: number, height: number, tileSize: number): number {
	if (tileSize < 1) {
		return 0;
	}

	let z = 0;
	while (levelPixels(width, z) > tileSize || levelPixels(height, z) > tileSize) {
		z++;
	}

	return z;
}





export function tileOrigin(tileSize: number, z: number, x: number): number {
	return x * tileSize * Math.pow(2, z);
}


export function tileSpan(tileSize: number, z: number): number {
	return tileSize * Math.pow(2, z);
}
